chips/rp2040/src/pwm.rs

```diff
@@ -0,0 +1,1254 @@
+        let default_config: PwmChannelConfiguration = PwmChannelConfiguration::default();
+        for channel_number in CHANNEL_NUMBERS {
+            self.configure_channel(channel_number, &default_config);
+            self.set_counter(channel_number, 0);
+            self.disable_interrupt(channel_number);
+        }
+        self.registers.intr.write(CH::CH.val(0));
+    }
+
+    // This method should be called when resolving dependencies for the
+    // default peripherals. See [crate::chip::Rp2040DefaultPeripherals::resolve_dependencies]
+    pub(crate) fn set_clocks(&self, clocks: &'a clocks::Clocks) {
+        self.clocks.set(clocks);
+    }
+
+    // Given a channel number and a channel pin, return a struct that allows controlling it
+    fn new_pwm_pin(&'a self, channel_number: ChannelNumber, channel_pin: ChannelPin) -> PwmPin<'a> {
+        PwmPin {
+            pwm_struct: self,
+            channel_number,
+            channel_pin,
+        }
+    }
+
+    // Map the given GPIO to a PWM channel and a PWM pin
+    fn gpio_to_pwm(&self, gpio: RPGpio) -> (ChannelNumber, ChannelPin) {
+        (ChannelNumber::from(gpio), ChannelPin::from(gpio))
+    }
+
+    /// Map the GPIO to a PwmPin struct
+    ///
+    /// The returned structure can be used to control the PWM pin.
+    ///
+    /// See [PwmPin]
+    pub fn gpio_to_pwm_pin(&'a self, gpio: RPGpio) -> PwmPin {
+        let (channel_number, channel_pin) = self.gpio_to_pwm(gpio);
+        self.new_pwm_pin(channel_number, channel_pin)
+    }
+
+    // Helper function to compute top, int and frac values
+    // selected_freq_hz ==> user's desired frequency
+    //
+    // Return value: Ok(top, int, frac) in case of no error, otherwise Err(())
+    fn compute_top_int_frac(&self, selected_freq_hz: usize) -> Result<(u16, u8, u8), ()> {
+        // If the selected frequency is high enough, then there is no need for a divider
+        // Note that unwrap can never fail.
+        let max_freq_hz = hil::pwm::Pwm::get_maximum_frequency_hz(self);
+        let threshold_freq_hz = max_freq_hz / hil::pwm::Pwm::get_maximum_duty_cycle(self);
+        if selected_freq_hz > threshold_freq_hz {
+            return Ok(((max_freq_hz / selected_freq_hz - 1) as u16, 1, 0));
+        }
+        // If the selected frequency is below the threshold frequency, then a divider is necessary
+
+        // Set top to max
+        let top = u16::MAX;
+        // Get the corresponding divider value
+        let divider = threshold_freq_hz as f32 / selected_freq_hz as f32;
+        // If the desired frequency is too low, then it can't be achieved using the divider.
+        // In this case, notify the caller with an error.
+        if divider >= 256.0f32 {
+            return Err(());
+        }
+        // At this point, the divider is a valid value. Its integral and fractional part
+        // can be computed.
+        let int = divider as u8;
+        let frac = ((divider - int as f32) * 16.0) as u8;
```

Comment:
These floating point calculations will be really slow/generate a lot of instructions since Tock does not currently support hardware floating point for arm cores. 

Fortunately, since you return an error if the divider is > 256, and only need 4-bit precision for the fractional component, I think the following should give the same results without requiring floating point:
```suggestion
        let divider = threshold_freq_hz / selected_freq_hz;
        // If the desired frequency is too low, then it can't be achieved using the divider.
        // In this case, notify the caller with an error.
        if divider >= 256.0f32 {
            return Err(());
        }
        // At this point, the divider is a valid value. Its integral and fractional part
        // can be computed.
        let int = divider as u8;
        // threshold_freq_hz *must* be less than u16::MAX since `get_maximum_duty_cycle()` 
        // returns u16::MAX + 1. Therefore neither of these multiplications by 16 can overflow.
        let frac = ((threshold_freq_hz * 16) / (selected_freq_hz) - divider * 16) as u8;
```

You may want to double check my math on this.

Reply:
Changed the code. Is it OK now?

Reply:
It is ok now once you make my other two suggested changes. I verified the new version is functionally equivalent to the old version using this playground https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3eaa2b45c63ca9a153ef00e2d49c8c6c (I had to run it locally to get past the playground timeout, though)